### PR TITLE
Data Catalog roots

### PR DESCRIPTION
**Pull request by savente93**
# Functional Requirements

## Solved
- [x] Data catalogs can not specify multiple roots to make the catalogs more portable and give better cross platform support as these can be either Windows or POSIX paths
- [x] Roots provided will be checked in order, first one to be found will be used. 
- [x] When none of the provided roots are found an error is raised 
- [x] if no roots are specified anywhere, the current working directory will be used. 

## Postponed
- [ ] Removing the other cross platform path transformations (Fixed in #788)
- [ ] Re enable Model Root logger file path handling  (Fixed in #788)

## To be agreed on
None so far that I am aware of

### CODE REVIEW

**Comment by savente93** [comment]
Fixes #721

File: data/catalogs/deltares_data.yml
Comment (DirkEilander) [inline data/catalogs/deltares_data.yml:7]: With these changes we should update the hydromt_version to >1.0 right? We should then also update the ref in the predefined catalogs.

**Review by DirkEilander — CHANGES_REQUESTED** [review]
Nice work @savente93. I have one major suggestion, see comment on the prioritisation of the `root` method argument vs the `yml['meta']['root']` value. Otherwise this looks good to me. If you can respond to that one I'll do another more thorough check. And perhaps good to ask a second reviewer.

File: hydromt/data_catalog.py
Comment (savente93) [reply hydromt/data_catalog.py:919]: Fair enough. I think I added this more as a testing as sometimes you call `from_dict` that isn't originating from a file (e.g. in testing) so this was meant as a fallback but I agree it has more potential for harm than good so I'll specify that in the tests if necessary .

File: data/catalogs/deltares_data.yml
Comment (savente93) [reply data/catalogs/deltares_data.yml:7]: sort of. You're right that the refs need to be updated, but only once this lands on main right? this one is going to v1 not main, and when we merge v1 to main, we'll have to update the refs anyway. Am I correct?

File: hydromt/data_catalog.py
Comment (savente93) [reply hydromt/data_catalog.py:835]: Good catch. This was from an earlier revision when `from_dict` and `from_yml` both needed to call `_determine_roots` and I think I forgot to remove it when that was no longer necessary. I'll remove it now

**Comment by savente93** [comment]
@DirkEilander should be good now, could you take another look?

File: hydromt/_validators/data_catalog.py
Comment (hboisgon) [inline hydromt/_validators/data_catalog.py:54]: I see in from_yml and from_dict that version can still be None. So should the hydromt_version here also be optional?

File: hydromt/data_catalog.py
Comment (hboisgon) [inline hydromt/data_catalog.py:814]: Not dirname(urlpath)?

File: tests/validators/test_data_catalog_validation.py
Comment (hboisgon) [inline tests/validators/test_data_catalog_validation.py:1]: Maybe add a test with from_yml and no root?

**Review by hboisgon — CHANGES_REQUESTED** [review]
Nice changes @savente93 and good we can finally support cross platform for the data catalogs :)
I just wonder if the case of from_yml with no root works but for the rest technically looks good to go!

I put a request changes because this time I won't approve if the changelog is not updated to reflect that multi platforms is supported, root became roots in yml and that now artifact_data kwargs is really removed. If we keep postponing we will forget things in the v1 changelog :)
Also I would like to already update the data catalog part of the documentation to update the change from root to roots and maybe a simple example to explain how this works because it sure is nice functionnality!

File: hydromt/_validators/data_catalog.py
Comment (savente93) [reply hydromt/_validators/data_catalog.py:54]: That version refers to the version of the catalogs, which it may or may not have, but `hydromt_version` is the version of hydromt that the catalog is compatible with. I think it is good to force people to specify what versions they are compatible with, since if they don't they won't get warnings if we make incompatible changes and use their old catalogs. What do you think?

File: hydromt/data_catalog.py
Comment (savente93) [reply hydromt/data_catalog.py:814]: Good catch.

File: tests/validators/test_data_catalog_validation.py
Comment (savente93) [reply tests/validators/test_data_catalog_validation.py:1]: sure, I'll add one.

File: data/catalogs/deltares_data.yml
Comment (savente93) [reply data/catalogs/deltares_data.yml:7]: You're right, it's probably best to update them now. I've updated all the catalog versions, but I do wonder if we still need to update the `predefined_catalogs.yml` since this hasn't landed on main yet, and I'm not sure how it will handle things that aren't on main yet. What do you two think? if so I'll make an issue to update those refs when we finsih v1

**Comment by savente93** [comment]
@hboisgon you make good points, thanks for that! I posed two questions in response to your review, when we can agree on those we should be good. in the mean time I'll add some documentation but if you both could take another look that'd be great (@DirkEilander cc)

**Comment by savente93** [comment]
@hboisgon I've updated the documentation. Do you think this is enough or will more be required?

File: data/catalogs/deltares_data.yml
Comment (hboisgon) [reply data/catalogs/deltares_data.yml]: True, forgot that the latest version points to main and only the others to a commit number. Is there a reason why actually we point to main and not a commit number for the latest version?
But yes it's not so easy to update so let's open the issue to update for later.

Comment (savente93) [reply data/catalogs/deltares_data.yml]: Why we point to main and not a specific commit I'm not sure. @DirkEilander knows most about that I think.

File: hydromt/_validators/data_catalog.py
Comment (hboisgon) [reply hydromt/_validators/data_catalog.py:54]: It's good to encourage it but not sure we want to force it. Especially since after v1 we hope the syntax and keywords of the catalog should be more stable so this might be a lot of trouble for the user to find out this information and the syntax and keep it updated with new hydromt versions. 
Maybe if None just print a warning that if the hydromt_version was not found, errors can happen later due to that. And do add the current used version? Then if the catalog is exported later it will print and can be re-used?
@DirkEilander what do you think?

File: docs/changelog.rst
Comment (hboisgon) [inline docs/changelog.rst:22]: Suggestion: maybe create a removed deprecation list in the v1 changelog rather than include them in the changed part? Also here we will list the --dd flag, staticmaps/staticgeoms etc.

File: docs/changelog.rst
Comment (hboisgon) [inline docs/changelog.rst:17]:
```suggestion
- Data catalogs can now list multiple version of the roots which will be searched for files depending on the system used (linux, windows etc). (#786)
```

File: docs/user_guide/data_prepare_cat.rst
Comment (hboisgon) [inline docs/user_guide/data_prepare_cat.rst:33]:
```suggestion
        - /linux/path/to/data_root/
```

File: docs/user_guide/data_prepare_cat.rst
Comment (hboisgon) [inline docs/user_guide/data_prepare_cat.rst:69]: Still optional in the case of cloud data for example. 
I would also explain that they actually link to the same folder and are for cross platform support only. Not like a data source per data source check will be done to be able to store data in different folders.

File: docs/user_guide/data_prepare_cat.rst
Comment (hboisgon) [inline docs/user_guide/data_prepare_cat.rst:73]:
```suggestion
  This can be used for cross platform and cross machine compatibility as can be seen above.
```

File: docs/user_guide/data_prepare_cat.rst
Comment (hboisgon) [inline docs/user_guide/data_prepare_cat.rst:75]: See the advice of Dirk but I have a preference to keep this as recommended

**Review by hboisgon — COMMENTED** [review]
Thanks @savente93 !
For the optional/mandatory hydromt_version, I personally prefer optional but let's see what @DirkEilander thinks.
Docs addition looks good too!
One thing to check that popped to mind is if for the cloud catalogs (no root), things still work and the data can be accessed and read. If so I think the rest is good :)

File: data/catalogs/aws_data.yml
Comment (savente93) [reply data/catalogs/aws_data.yml:4]: good point, I'll double check that I wrote tests for that

Comment (savente93) [reply data/catalogs/aws_data.yml:4]: it turns out I didn't and there were some troubles there. I fixed them now but good catch!

File: docs/changelog.rst
Comment (savente93) [reply docs/changelog.rst:22]: sure, good idea!

File: hydromt/_validators/data_catalog.py
Comment (savente93) [reply hydromt/_validators/data_catalog.py:54]: I'm okay with this solution, I'll go with that

File: data/catalogs/artifact_data.yml
Comment (hboisgon) [inline data/catalogs/artifact_data.yml:269]: This is actually issue #696 and I think we should still fix it for main as well. If you need this for tests to pass then we can keep and solve the merge conflicts with main when they come :)

File: hydromt/_validators/data_catalog.py
Comment (hboisgon) [inline hydromt/_validators/data_catalog.py:66]:
```suggestion
                f"No hydromt version was specified for the data catalog, thus compatibility between used hydromt version ({HYDROMT_VERSION}) and the catalog could not be determined."
```

File: tests/validators/test_data_catalog_validation.py
Comment (hboisgon) [inline tests/validators/test_data_catalog_validation.py:1]: Are these two tests still run in the current test bench? Just to be extra sure the data can be read as well:

https://github.com/Deltares/hydromt/blob/d634534fd1c3c5e470d6835a1ee1c518705576ad/tests/test_data_adapter.py#L98-L126

**Review by hboisgon — APPROVED** [review]
Great! Looks good now :)
Just a final couple of comments/typos but else looks good to go.

File: data/catalogs/artifact_data.yml
Comment (savente93) [reply data/catalogs/artifact_data.yml:269]: They are indeed necessary for these tests. I'll open a separate PR to fix these in main if you want.

File: tests/validators/test_data_catalog_validation.py
Comment (savente93) [reply tests/validators/test_data_catalog_validation.py:1]: I'm not quite sure I understand what the question is, but yes, the tests are run against the files on this branch, though they only test the syntax of the file, not the actually data as the test you linked seems to do

File: data/catalogs/artifact_data.yml
Comment (hboisgon) [reply data/catalogs/artifact_data.yml:269]: I think @Tjalling-dejong will or has opened a larger one to update artifact_data catalog (also driver_kwargs should be updated I think)

File: tests/validators/test_data_catalog_validation.py
Comment (hboisgon) [reply tests/validators/test_data_catalog_validation.py:1]: If in the current v1 tests, these are included and pass or if they are skipped for some reason. If they are not skipped and pass, then we are sure the data can actually still successfully be read on top of the extra test you implemented in this PR

Comment (savente93) [reply tests/validators/test_data_catalog_validation.py:1]: they should be included yes

File: tests/validators/test_data_catalog_validation.py
Comment (DirkEilander) [inline tests/validators/test_data_catalog_validation.py:157]: nice one! didn't know it was possible to test the logs like this

**Review by DirkEilander — APPROVED** [review]
Looks great @savente93!
I've just added a few comments for clarification in the changelog, removing an error message and adding a root to deltares_data.yml. With these changes the PR can be merged I think.

File: tests/validators/test_data_catalog_validation.py
Comment (savente93) [reply tests/validators/test_data_catalog_validation.py:25]: good one, forgot about that

**Pull request merged**
